drivers/bus/acpi/eval.c

```diff
@@ -1,255 +1,804 @@
-             if (Obj->Buffer.Length > sizeof(ULONG))
-                 ExtraParamLength += Obj->Buffer.Length + 1 - sizeof(ULONG);
-             break;
-
-          case ACPI_TYPE_PACKAGE:
-             Status = GetPackageSize(Obj, &Count, &ExtraParamLength);
-             if (!NT_SUCCESS(Status))
-                 return Status;
-             break;
-
-          default:
-             ASSERT(FALSE);
-             return STATUS_UNSUCCESSFUL;
-      }
-
-      DPRINT("ExtraParamLength %lu\n", ExtraParamLength);
-      OutputBuf = ExAllocatePoolWithTag(NonPagedPool,
-                                        sizeof(ACPI_EVAL_OUTPUT_BUFFER) - sizeof(ACPI_METHOD_ARGUMENT) + ExtraParamLength,
-                                        'BpcA');
-      if (!OutputBuf)
-          return STATUS_INSUFFICIENT_RESOURCES;
-
-      OutputBuf->Signature = ACPI_EVAL_OUTPUT_BUFFER_SIGNATURE;
-      OutputBuf->Length = ExtraParamLength;
-      OutputBuf->Count = Count;
-
-      switch (Obj->Type)
-      {
-          case ACPI_TYPE_INTEGER:
-             ACPI_METHOD_SET_ARGUMENT_INTEGER(OutputBuf->Argument, Obj->Integer.Value);
-             break;
-
-          case ACPI_TYPE_STRING:
-             ACPI_METHOD_SET_ARGUMENT_STRING(OutputBuf->Argument, Obj->String.Pointer);
-             break;
-
-          case ACPI_TYPE_BUFFER:
-             ACPI_METHOD_SET_ARGUMENT_BUFFER(OutputBuf->Argument, Obj->Buffer.Pointer, Obj->Buffer.Length);
-             break;
-
-          case ACPI_TYPE_PACKAGE:
-             Status = ConvertPackageArguments(OutputBuf->Argument, Obj);
-             if (!NT_SUCCESS(Status))
-                 return Status;
-             break;
-
-          default:
-             ASSERT(FALSE);
-             return STATUS_UNSUCCESSFUL;
-      }
-
-      if (IrpSp->Parameters.DeviceIoControl.OutputBufferLength >= sizeof(ACPI_EVAL_OUTPUT_BUFFER) - sizeof(ACPI_METHOD_ARGUMENT) +
-                                                                  ExtraParamLength)
-      {
-          RtlCopyMemory(Irp->AssociatedIrp.SystemBuffer, OutputBuf, sizeof(ACPI_EVAL_OUTPUT_BUFFER) - sizeof(ACPI_METHOD_ARGUMENT) +
-                                                                    ExtraParamLength);
-          Irp->IoStatus.Information = sizeof(ACPI_EVAL_OUTPUT_BUFFER) - sizeof(ACPI_METHOD_ARGUMENT) + ExtraParamLength;
-          ExFreePoolWithTag(OutputBuf, 'BpcA');
-          return STATUS_SUCCESS;
-      }
-      else
-      {
-          ExFreePoolWithTag(OutputBuf, 'BpcA');
-          return STATUS_BUFFER_TOO_SMALL;
-      }
-  }
-  else
-  {
-      DPRINT1("Query method %4s failed on %p\n", EvalInputBuff->MethodName, DeviceData->AcpiHandle);
-      return STATUS_UNSUCCESSFUL;
-  }
+    PEVAL_WORKITEM_DATA WorkItemData = Parameter;
+    NTSTATUS Status;
+    PIRP Irp;
+
+    PAGED_CODE();
+
+    Irp = WorkItemData->Irp;
+
+    Status = Bus_PDO_EvalMethod(WorkItemData->DeviceData, Irp);
+
+    ExFreePoolWithTag(WorkItemData, 'ipcA');
```

Comment:
We don't have a define TAG_.... for acpi?

Reply:
All allocation tags are currently hardcoded.